util/global_conn_id.go

```diff
@@ -0,0 +1,453 @@
+	g.serverIDGetter = serverIDGetter
+
+	g.local32.Init(nil)
+	g.local64.Init(existedChecker)
+
+	g.is64bits.Set(1) // TODO: set 32bits as default, after 32bits logics is fully implemented and tested.
+}
+
+// SetServerIDGetter set serverIDGetter member.
+func (g *GlobalConnIDAllocator) SetServerIDGetter(serverIDGetter serverIDGetterFn) {
+	g.serverIDGetter = serverIDGetter
+}
+
+// NextID returns next connection ID.
+func (g *GlobalConnIDAllocator) NextID() uint64 {
+	globalConnID := g.Allocate()
+	return globalConnID.ID()
+}
+
+// Allocate allocates a new global connID.
+func (g *GlobalConnIDAllocator) Allocate() GlobalConnID {
+	serverID := g.serverIDGetter()
+
+	// 32bits.
+	if !g.Is64() {
+		localConnID, ok := g.local32.Allocate()
+		if ok {
+			return GlobalConnID{
+				ServerID:    serverID,
+				LocalConnID: localConnID,
+				Is64bits:    false,
+			}
+		}
+		g.UpgradeTo64() // go on to 64bits.
+	}
+
+	// 64bits.
+	return GlobalConnID{
+		ServerID:    serverID,
+		LocalConnID: g.local64.Allocate(serverID),
+		Is64bits:    true,
+	}
+}
+
+// Release releases connectionID to pool.
+func (g *GlobalConnIDAllocator) Release(connectionID uint64) {
+	globalConnID, isTruncated, err := ParseGlobalConnID(connectionID)
+	if err != nil || isTruncated {
+		logutil.BgLogger().Error("failed to ParseGlobalConnID", zap.Error(err), zap.Uint64("connectionID", connectionID), zap.Bool("isTruncated", isTruncated))
+		return
+	}
+
+	if globalConnID.Is64bits {
+		g.local64.Deallocate(globalConnID.LocalConnID)
+	} else {
+		if err = g.local32.Deallocate(globalConnID.LocalConnID); err != nil {
+			logutil.BgLogger().Error("failed to release 32bits connection ID", zap.Error(err), zap.Uint64("connectionID", connectionID), zap.Uint64("localConnID", globalConnID.LocalConnID))
+		}
+	}
+}
+
+// LocalConnIDAllocator64 is local connID allocator for 64bits global connection ID.
+type LocalConnIDAllocator64 struct {
+	existedChecker connectionIDExistCheckerFn
+	lastID         uint64
+}
+
+// Init initiates LocalConnIDAllocator64
+func (a *LocalConnIDAllocator64) Init(existedChecker connectionIDExistCheckerFn) {
+	a.existedChecker = existedChecker
+}
+
+// LocalConnIDAllocator64RetryCount is the retry count of `LocalConnIDAllocator64.Allocate`
+const LocalConnIDAllocator64RetryCount = 20
+
+// Allocate local connID for 64bits global connID.
+// local connID with 40bits pool size is big enough and should not be exhausted, as `MaxServerConnections` is a uint32.
+func (a *LocalConnIDAllocator64) Allocate(serverID uint64) (localConnID uint64) {
+	for i := 0; i < LocalConnIDAllocator64RetryCount; i++ {
+		localConnID := atomic.AddUint64(&a.lastID, 1) & MaxLocalConnID64
+		if !a.existedChecker(makeGlobalConnID(true, serverID, localConnID)) {
+			return localConnID
+		}
+	}
+	panic(fmt.Sprintf("Failed to allocate 64bits local connID after retry %v times. Should never happen", LocalConnIDAllocator64RetryCount))
```

Comment:
It might not be never happen (from the perspective of the allocator itself, since a caller can set a relatively small ID range). How about throwing errors and let caller to decide what to do when allocate failed?